Mathlib/Topology/MetricSpace/Congruence.lean

```diff
@@ -0,0 +1,219 @@
+
+## Notation
+
+* `v₁ ≅ v₂`: for `Congruent v₁ v₂`.
+-/
+
+variable {ι ι' : Type*} {P₁ P₂ P₃ : Type*} {v₁ : ι → P₁} {v₂ : ι → P₂} {v₃ : ι → P₃}
+
+noncomputable section
+
+/-- A congruence between indexed sets of vertices v₁ and v₂.
+Use `open scoped Congruence` to access the `v₁ ≅ v₂` notation. -/
+def Congruent (v₁ : ι → P₁) (v₂ : ι → P₂) [PseudoEMetricSpace P₁]
+    [PseudoEMetricSpace P₂] : Prop :=
+  ∀ (i₁ i₂ : ι), (edist (v₁ i₁) (v₁ i₂) = edist (v₂ i₁) (v₂ i₂))
+
+@[inherit_doc]
+scoped[Congruence] infixl:25 " ≅ " => Congruent
+
+/-- A congruence holds if and only if all extended distances are the same. -/
+lemma congruent_iff_edist_eq [PseudoEMetricSpace P₁] [PseudoEMetricSpace P₂] :
+    Congruent v₁ v₂ ↔ (∀ (i₁ i₂ : ι), (edist (v₁ i₁) (v₁ i₂) = edist (v₂ i₁) (v₂ i₂))) :=
+  forall₂_congr (fun _ _ => by simp only [edist])
+
+/-- A congruence holds if and only if all non-negative distances are the same. -/
+lemma congruent_iff_nndist_eq [PseudoMetricSpace P₁] [PseudoMetricSpace P₂] :
+    Congruent v₁ v₂ ↔ (∀ (i₁ i₂ : ι), (nndist (v₁ i₁) (v₁ i₂) = nndist (v₂ i₁) (v₂ i₂))) :=
+  forall₂_congr (fun _ _ => by rw [edist_nndist, edist_nndist]; norm_cast )
+
+/-- A congruence holds if and only if all distances are the same. -/
+lemma congruent_iff_dist_eq [PseudoMetricSpace P₁] [PseudoMetricSpace P₂] :
+    Congruent v₁ v₂ ↔ (∀ (i₁ i₂ : ι), (dist (v₁ i₁) (v₁ i₂) = dist (v₂ i₁) (v₂ i₂))) :=
+  congruent_iff_nndist_eq.trans
+    (forall₂_congr (fun _ _ => by rw [dist_nndist, dist_nndist]; norm_cast))
+
+
+namespace Congruence
+
+/-- A congruence preserves extended distance. -/
+alias ⟨edist_eq, _⟩ := congruent_iff_edist_eq
+
+/-- A congruence follows from preserved extended distance. -/
+alias ⟨_, of_edist_eq⟩ := congruent_iff_edist_eq
+
+/-- A congruence preserves non-negative distance. -/
+alias ⟨nndist_eq, _⟩ := congruent_iff_nndist_eq
+
+/-- A congruence follows from preserved non-negative distance. -/
+alias ⟨_, of_nndist_eq⟩ := congruent_iff_nndist_eq
+
+/-- A congruence preserves distance. -/
+alias ⟨dist_eq, _⟩ := congruent_iff_dist_eq
+
+/-- A congruence follows from preserved distance. -/
+alias ⟨_, of_dist_eq⟩ := congruent_iff_dist_eq
+
+/-- A congruence follows from pairwise preserved extended distance. -/
+lemma of_pairwise_edist_eq [PseudoEMetricSpace P₁] [PseudoEMetricSpace P₂] [DecidableEq ι]
+    (h : Pairwise (fun i₁ i₂ => (edist (v₁ i₁) (v₁ i₂) =
+      edist (v₂ i₁) (v₂ i₂)))) : v₁ ≅ v₂ :=
+  fun i₁ i₂ => if g : i₁ = i₂ then by rw [g]; simp else h g
+
+/-- A congruence follows from pairwise preserved non-negative distance. -/
+lemma of_pairwise_nndist_eq [PseudoMetricSpace P₁] [PseudoMetricSpace P₂] [DecidableEq ι]
+    (h : Pairwise (fun i₁ i₂ => (nndist (v₁ i₁) (v₁ i₂) = nndist (v₂ i₁) (v₂ i₂)))) : v₁ ≅ v₂ :=
+  of_pairwise_edist_eq (fun i₁ i₂ hn => by
+    simp only [edist_nndist]
+    norm_cast
+    exact h hn)
```

Comment:
Here you can probably use `exact_mod_cast h hn`. Again this is not a request for change, only information for you.